sdk/eventhub/Azure.Messaging.EventHubs/src/Processor/EventProcessor.cs

```diff
@@ -0,0 +1,231 @@
+        /// </summary>
+        ///
+        private ConcurrentDictionary<string, PartitionPump> PartitionPumps { get; }
+
+        /// <summary>
+        ///   The running task responsible for checking the status of the owned partition pumps.
+        /// </summary>
+        ///
+        private Task RunningTask { get; set; }
+
+        /// <summary>
+        ///   Initializes a new instance of the <see cref="EventProcessor"/> class.
+        /// </summary>
+        ///
+        /// <param name="consumerGroup">The name of the consumer group this event processor is associated with.  Events are read in the context of this group.</param>
+        /// <param name="eventHubClient">The client used to interact with the Azure Event Hubs service.</param>
+        /// <param name="partitionProcessorFactory">A factory used to create partition processors.  Its implementation must be provided by the user.</param>
+        /// <param name="partitionManager">A partition manager used to interact with the storage system.  Its implementation must be provided by the user.</param>
+        /// <param name="options">The set of options to use for this event processor.</param>
+        ///
+        public EventProcessor(string consumerGroup,
+                              EventHubClient eventHubClient,
+                              IPartitionProcessorFactory partitionProcessorFactory,
+                              IPartitionManager partitionManager,
+                              EventProcessorOptions options = default)
+        {
+            Guard.ArgumentNotNull(nameof(eventHubClient), eventHubClient);
+            Guard.ArgumentNotNullOrEmpty(nameof(consumerGroup), consumerGroup);
+            Guard.ArgumentNotNull(nameof(partitionProcessorFactory), partitionProcessorFactory);
+            Guard.ArgumentNotNull(nameof(partitionManager), partitionManager);
+
+            InnerClient = eventHubClient;
+            ConsumerGroup = consumerGroup;
+            PartitionProcessorFactory = partitionProcessorFactory;
+            PartitionManager = partitionManager;
+            Options = options?.Clone() ?? new EventProcessorOptions();
+
+            Identifier = Guid.NewGuid().ToString();
+            PartitionPumps = new ConcurrentDictionary<string, PartitionPump>();
+        }
+
+        /// <summary>
+        ///   Starts the event processor.  In case it's already running, nothing happens.
+        /// </summary>
+        ///
+        /// <returns>A task to be resolved on when the operation has completed.</returns>
+        ///
+        public async Task StartAsync()
+        {
+            if (RunningTask == null)
+            {
+                await RunningTaskSemaphore.WaitAsync().ConfigureAwait(false);
+
+                try
+                {
+                    if (RunningTask == null)
+                    {
+                        RunningTaskTokenSource = new CancellationTokenSource();
```

Comment:
nit:  Just to be sure, I'd recommend cancelling any token source that currently exists.   `RunningTaskTokenSource?.Cancel()`